File: src/main/java/com/jetbrains/idear/asr/GrammarService.java
package com.jetbrains.idear.asr;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.actionSystem.ActionManager;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.actionSystem.ex.AnActionListener;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.util.Disposer;


/**
 * Created by breandan on 7/10/2015.
 */
public class GrammarService implements Disposable {
    private static final Logger logger = Logger.getInstance(GrammarService.class);

    @Override
    public void dispose() {
        Disposer.dispose(this);
    }

    public void init() {
        ActionManager actionManager = ActionManager.getInstance();

        actionManager.addAnActionListener(new AnActionListener() {
            @Override
            public void beforeActionPerformed(AnAction anAction, DataContext dataContext, AnActionEvent anActionEvent) {
                String actionId = actionManager.getId(anAction);

                if ("someaction".equals(actionId)) {
                    logger.info("Swapping in grammar for action: " + anAction.toString());
                    //swap in a context dependent grammar
                }
            }

            @Override
            public void afterActionPerformed(AnAction anAction, DataContext dataContext, AnActionEvent anActionEvent) {

                //swap out a context dependent grammar
            }

            @Override
            public void beforeEditorTyping(char c, DataContext dataContext) {

            }
        });
    }
}

File: src/main/java/org/openasr/idear/asr/GrammarService.kt
package org.openasr.idear.asr

import com.intellij.openapi.actionSystem.*
import com.intellij.openapi.actionSystem.ex.AnActionListener
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.diagnostic.logger
import com.intellij.util.messages.MessageBus
import org.openasr.idear.utils.ActionUtils

object GrammarService : AnActionListener {
    private val log = logger<GrammarService>()

    fun init(bus: MessageBus) = bus.connect().subscribe(AnActionListener.TOPIC, this)

    fun useCommandGrammar() {
        val grammar = HashSet<String>()

        addIdearGrammar(grammar)
        ActionUtils.addActionWords(grammar)

        AsrService.setGrammar(grammar.toTypedArray())
    }

    fun useDictationGrammar() = AsrService.setGrammar(emptyArray())

    private fun addIdearGrammar(grammar: HashSet<String>) =
        grammar.addAll(arrayOf("command", "dictation", "mode"))

    override fun beforeActionPerformed(anAction: AnAction,
                                       anActionEvent: AnActionEvent) {
        val actionId = ActionManager.getInstance().getId(anAction)

        if ("someaction" == actionId) {
            log.info("Swapping in grammar for action: $anAction")
            //swap in a context aware grammar
        }
    }

    override fun beforeEditorTyping(c: Char, dataContext: DataContext) = Unit
}
